[FLINK-5654] - Add processing time OVER RANGE BETWEEN x PRECEDING aggregation to SQL

rtudoran opened the pull request:
Added tests based on harness util framework

Thanks for contributing to Apache Flink. Before you open your pull request, please take the following check list into consideration.
If your changes take all of the items into account, feel free to open your pull request. For more information and/or questions please refer to the [How To Contribute guide](http://flink.apache.org/how-to-contribute.html).
In addition to going through the list, please provide a meaningful description of your changes.

- [x ] General
  - The pull request references the related JIRA issue ("[FLINK-XXX] Jira title text")
  - The pull request addresses only one issue
  - Each commit in the PR has a meaningful commit message (including the JIRA id)

- [x ] Documentation
  - Documentation has been added for new functionality
  - Old documentation affected by the pull request has been updated
  - JavaDoc for public methods has been added

- [x ] Tests & Build
  - Functionality added by the pull request is covered by tests
  - `mvn clean verify` has been executed successfully locally or a Travis build has passed

rtudoran commented:
@sunjincheng121 @fhueske 
In order to revert the mess after the rebase that contained other comits, i have reopened the PR.
I will close the previous one. Please consider this. It is the same

sunjincheng121 commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/plan/nodes/datastream/DataStreamOverAggregate.scala:
```diff
@@ -17,24 +17,38 @@
+
+import org.apache.flink.table.runtime.aggregate.AggregateUtil.CalcitePair
+import org.apache.calcite.plan.RelOptCluster
+import org.apache.calcite.plan.RelTraitSet
```
Please remove un-use import.

sunjincheng121 commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/plan/nodes/datastream/DataStreamOverAggregate.scala:
```diff
@@ -17,24 +17,38 @@
+import org.apache.flink.streaming.api.windowing.time.Time
+import org.apache.flink.api.java.tuple.Tuple
+import org.apache.flink.util.Collector
+import java.util.concurrent.TimeUnit
```
Please remove un-use import.

sunjincheng121 commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/plan/nodes/datastream/DataStreamOverAggregate.scala:
```diff
@@ -274,6 +286,57 @@ class DataStreamOverAggregate(
+        // partitioned aggregation
+        if (partitionKeys.nonEmpty) {
+          
+          val processFunction = AggregateUtil.createTimeBoundedProcessingOverProcessFunction(
```
Move it to the top of `if` statement.

sunjincheng121 commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/plan/nodes/datastream/DataStreamOverAggregate.scala:
```diff
@@ -274,6 +286,57 @@ class DataStreamOverAggregate(
+          .name(aggOpName)
+          .asInstanceOf[DataStream[Row]]
+        } else { // non-partitioned aggregation
+          val processFunction = AggregateUtil.createTimeBoundedProcessingOverProcessFunction(
```
We can remove it If we have created processFunction before if statement.

rtudoran replied:
@sunjincheng121 I disagree with this. Even if we move the creation of the process function at the top of if (which is fine), we still need to make the differentiation on what to use in the KeyBy - the actual field or a NullSelector. Hence, as the if is needed i prefer to keep it as it is

fhueske replied:
The `isPartitioned` parameter is not used in the `createTimeBoundedProcessingOverProcessFunction()`. So both calls are identical and we can move it outside of the condition and save a few LOCs.

rtudoran replied:
ok... :)

sunjincheng121 commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/AggregateUtil.scala:
```diff
@@ -1208,5 +1208,40 @@ object AggregateUtil {
+    namedAggregates: Seq[CalcitePair[AggregateCall, String]],
+    inputType: RelDataType,
+    timeBoundary: Long,
+    isPartitioned: Boolean = true): ProcessFunction[Row, Row] = {
```
We can remove this param. What do you think?

fhueske replied:
Yes, it is not used in the function. Removing it should be fine.

sunjincheng121 commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/AggregateUtil.scala:
```diff
@@ -1208,5 +1208,40 @@ object AggregateUtil {
+  private[flink] def createTimeBoundedProcessingOverProcessFunction(
+    namedAggregates: Seq[CalcitePair[AggregateCall, String]],
+    inputType: RelDataType,
+    timeBoundary: Long,
```
Just a suggestion:  `timeBoundary -> precedingOffset`. :)

rtudoran replied:
@sunjincheng121  :) ...i feel it is more expressive to mention time as the window is related to time boundaries. Offset for me sounds more related to indexes...but i am not a native speaker though...

sunjincheng121 commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/ProcTimeBoundedProcessingOverProcessFunction.scala:
```diff
@@ -0,0 +1,169 @@
+import org.apache.flink.api.common.typeinfo.BasicTypeInfo
+import org.apache.flink.api.java.typeutils.TupleTypeInfo
+import scala.collection.mutable.Queue
+import org.apache.flink.api.common.typeinfo.TypeHint
```
Please remove the un-use import.

sunjincheng121 commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/ProcTimeBoundedProcessingOverProcessFunction.scala:
```diff
@@ -0,0 +1,169 @@
+import java.util.{ ArrayList, LinkedList, List => JList }
+
+/**
+ * Process Function used for the aggregate in partitioned bounded windows in
```
Remove `partitioned `. 
`Process Function used for the aggregate in partitioned bounded` ->` Process Function used for the aggregate in bounded proc-time OVER window`

sunjincheng121 commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/ProcTimeBoundedProcessingOverProcessFunction.scala:
```diff
@@ -0,0 +1,169 @@
+  private val aggFields: Array[Int],
+  private val forwardedFieldCount: Int,
+  private val rowTypeInfo: RowTypeInfo,
+  private val timeBoundary: Long,
```
Just a suggestion:  `timeBoundary -> precedingOffset`. :)

rtudoran replied:
see previous comment

fhueske replied:
Another option would be `precedingTimeBoundary` ;-). But I'm also fine with `timeBoundary`.

rtudoran replied:
precedingTimeBoundary seems to be the combination of the 3 opinions :)

sunjincheng121 commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/ProcTimeBoundedProcessingOverProcessFunction.scala:
```diff
@@ -0,0 +1,169 @@
+    val limit = currentTime - timeBoundary
+    
+    // we iterate through all elements in the window buffer based on timestampt keys
+    // when we fine timestamps that are out of interest, we need to get the corresponding elements
```
`fine` -> `find`

sunjincheng121 commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/ProcTimeBoundedProcessingOverProcessFunction.scala:
```diff
@@ -0,0 +1,169 @@
+    }
+
+    //update the elements to be removed and retract them from aggregators
+    val limit = currentTime - timeBoundary
```
Just a suggestion: `limit -> smallestTs`. What do you think.

rtudoran replied:
@sunjincheng121 for me it is more intuitive to think of this as a limit as it defines the threshold...but again as i am not native speaker ...

sunjincheng121 commented on flink-libraries/flink-table/src/test/scala/org/apache/flink/table/api/scala/stream/sql/SqlITCase.scala:
```diff
@@ -19,7 +19,10 @@
+import org.apache.flink.table.api.scala.stream.sql.SqlITCase.{
+  EventTimeSourceFunction, 
+  RowResultSortComparator,
+  TupleRowSelector}
```
Keep in one line.

rtudoran replied:
Done...but i must admit i am confused with the application of the style. I thought we go for lines of less than 100 characters...why would this be an exception?

@sunjincheng121

sunjincheng121 commented on flink-libraries/flink-table/src/test/scala/org/apache/flink/table/api/scala/stream/sql/SqlITCase.scala:
```diff
@@ -696,6 +716,205 @@ class SqlITCase extends StreamingWithStateTestBase {
+  
+  @Test
+  def testCountAggregatationProcTimeHarnessPartitioned(): Unit = {
+    val env = StreamExecutionEnvironment.getExecutionEnvironment
```
Can you explain why we need this test?

rtudoran replied:
Simple answer is that @fhueske  asked for it for quite some time
Long answer is that we need to use the framework that was developed specifically to check the proctime working of the operators.

fhueske replied:
Yes, the test harness allows to "control" the processing time (basically the time that this returned if you ask the process function for the current processing time).
I would move this test to a separate file though because it is a unit test (class name should end with `Test`) and is a test specifically for the `ProcTimeBoundedProcessingOverProcessFunction`.

sunjincheng121 replied:
Thanks for the explanation. +1

sunjincheng121 commented on flink-libraries/flink-table/pom.xml:
```diff
@@ -140,6 +140,20 @@ under the License.
+			<version>${project.version}</version>
+			<scope>test</scope>
+			<type>test-jar</type>
+		</dependency>
```
Can we remove this dependency If we remove `testCountAggregatationProcTimeHarnessPartitioned`?

rtudoran replied:
@sunjincheng121  - as per my previous answer - @fhueske  asked explicitly for this test and dependencies. Initially there was suppose to be only one dependency but this had an inner dependency and that is why you have these.

fhueske replied:
Yes, the dependency is required for the test harness class which allows to control processing time. This is needed to properly test processing time operators that have a semantic that depends on the time (so not only sorting by time which is implicitly true for processing time). The processing time OVER RANGE window groups data by processing time. If we want to test this without manual timing (which is not good unit test practice) we need the test harness. There will be more operators in the future that require this dependency.

sunjincheng121 replied:
thanks for explanation. @rtudoran @fhueske

rtudoran commented:
@sunjincheng121 - thanks for the review. I addressed most of the comments. 
@fhueske

fhueske commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/AggregateUtil.scala:
```diff
@@ -1208,5 +1208,40 @@ object AggregateUtil {
+      transformToAggregateFunctions(
+        namedAggregates.map(_.getKey),
+        inputType,
+        needRetraction = false)
```
`needRetraction` must be `true`. This is probably not caught by the tests, because the agg functions in the test are always retractable. You could use a `min` or `max` aggregation which have dedicated implementation for retraction. 

Actually, we should check if the over window tests have `min` and `max` agg functions as well. I'll do that once all over windows have been merged.

rtudoran replied:
OK...i modified this. However i am not sure i perfectly understand the impact of this parameter in this case of proctime.

fhueske replied:
The parameter asks for AggregateFunctions that support retraction. There are some functions (like min and max) that are less efficient if they need to prepare for retraction. Therefore they have a retractable and a non-retractable version. The parameter decides which version is returned.

fhueske commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/AggregateUtil.scala:
```diff
@@ -1208,5 +1208,40 @@ object AggregateUtil {
+        needRetraction = false)
 
+    val aggregationStateType: RowTypeInfo =
+      createDataSetAggregateBufferDataType(Array(), aggregates, inputType)
```
use `createAccumulatorRowType(aggregates)`

fhueske commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/ProcTimeBoundedProcessingOverProcessFunction.scala:
```diff
@@ -0,0 +1,166 @@
+  override def open(config: Configuration) {
+    output = new Row(forwardedFieldCount + aggregates.length)
+
+    // We keep the elements received in a list state 
```
update comment? We keep the data in a `MapState` indexed by ingestion time?

rtudoran replied:
ok

fhueske commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/ProcTimeBoundedProcessingOverProcessFunction.scala:
```diff
@@ -0,0 +1,166 @@
+    //update the elements to be removed and retract them from aggregators
+    val limit = currentTime - timeBoundary
+    
+    // we iterate through all elements in the window buffer based on timestampt keys
```
-t -> timestamp

rtudoran replied:
this was tricky :) i was looking at the math formula and try to figure out what -t would imply :)

fhueske commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/ProcTimeBoundedProcessingOverProcessFunction.scala:
```diff
@@ -0,0 +1,166 @@
+      val elementKey = iter.next
+      if (elementKey < limit) {
+        val elementsRemove = rowMapState.get(elementKey)
+        val iterRemove = elementsRemove.iterator()
```
the returned list is an ArrayList. Can we iterate over it with an index and not with an Iterator?

rtudoran replied:
@fhueske is there a difference? i would expect that the iterator is also implemented via an index that is implemented. Or perhaps you are after reducing the number of objects that are created?

fhueske replied:
We would save the `Iterator` object instance. Probably micro optimization, but since we iterate most lists with `while(i < x)` we could do the same here.

fhueske commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/ProcTimeBoundedProcessingOverProcessFunction.scala:
```diff
@@ -0,0 +1,166 @@
+        val elementsRemove = rowMapState.get(elementKey)
+        val iterRemove = elementsRemove.iterator()
+        while (iterRemove.hasNext()) {
+         val remove = iterRemove.next() 
```
indent by 2 chars

fhueske commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/ProcTimeBoundedProcessingOverProcessFunction.scala:
```diff
@@ -0,0 +1,166 @@
+      }
+    }
+    //need to remove in 2 steps not to have concurrent access errors via iterator to the MapState
+    var iterRemove = markToRemove.iterator()
```
iterate by index and not `Iterator`

rtudoran replied:
fine...but same question as before - is it more efficient for some reason?

fhueske commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/ProcTimeBoundedProcessingOverProcessFunction.scala:
```diff
@@ -0,0 +1,166 @@
+    // we iterate through all elements in the window buffer based on timestampt keys
+    // when we find timestamps that are out of interest, we need to get the corresponding elements
+    // and eliminate them. Multiple elements can be received at the same timestamp
+    val iter = rowMapState.keys.iterator
```
We could add another `ValueState` to remember the highest key that had been removed. If we have this information, we only need to iterate once per millisecond over the map keys. Would help if multiple rows arrive in the same millisecond.

rtudoran replied:
I like the idea. I think it can help to reduce the search for elements to be removed!
Thanks
@fhueske

fhueske commented on flink-libraries/flink-table/src/test/scala/org/apache/flink/table/api/scala/stream/sql/SqlITCase.scala:
```diff
@@ -696,6 +713,205 @@ class SqlITCase extends StreamingWithStateTestBase {
+  @Test
+  def testCountAggregatationProcTimeHarnessPartitioned(): Unit = {
+    val env = StreamExecutionEnvironment.getExecutionEnvironment
+    val tEnv = TableEnvironment.getTableEnvironment(env)
```
The test harness test does not need `StreamExecutionEnvironment` and `TableEnvironment`.

fhueske commented on flink-libraries/flink-table/src/test/scala/org/apache/flink/table/api/scala/stream/sql/SqlITCase.scala:
```diff
@@ -696,6 +713,205 @@ class SqlITCase extends StreamingWithStateTestBase {
+      rOutput.setField(3, "aaa")
+      rOutput.setField(4, 11L)
+      rOutput.setField(5, 1L)   //count is 1
+    expectedOutput.add(new StreamRecord(rOutput, 1001));
```
please remove `;` (also in the following lines)

rtudoran replied:
:) beautiful java trademark

sunjincheng121 commented on flink-libraries/flink-table/src/test/scala/org/apache/flink/table/api/scala/stream/sql/SqlITCase.scala:
```diff
@@ -696,6 +713,205 @@ class SqlITCase extends StreamingWithStateTestBase {
+
+  
+  @Test
+  def testCountAggregatationProcTimeHarnessPartitioned(): Unit = {
```
`Aggregatation` -> `Aggregation` ?

rtudoran commented:
@sunjincheng121 @fhueske i have implemented all the changes you mentioned. 
-moved to index list traversals from iterators
-keep last proctime in a valuestate to avoid searching for elements to be removed several times in case events arrive within the same timestamp
-i moved the harness test in astandalone class name proctimetest in order to support potential future tests that would do proctime verifications

fhueske commented:
Thanks for the update @rtudoran.
Will merge this :-)

fhueske commented:
Oh, I just remembered that the OVER RANGE semantics are slightly different. All rows that arrive in the same millisecond need to get the same aggregation value. So we need not only to retract all records which are too old but also accumulate all records which are received in the same millisecond.

Therefore, we would need to redesign the `ProcessFunction` a bit.

- In `processElement()` we put the new record in the MapState and register a processing time timer for current time + 1. This will create a call back on `onTimer()` when current time + 1 is reached. 
- When `onTimer()` is called, we process the rows of timestamp - 1, retract all old values accumulate all new values and emit all rows of timestamp - 1. The implementation of `onTimer()` can reuse most of what is currently done in `processElement()`. 

Does that make sense @rtudoran? 
Do you want to make the change? Otherwise, I can also do it before merging. 

Sorry for recognizing this just now :-/

Best, Fabian

rtudoran commented:
@fhueske 
Thanks for the feedback - i can of course do the modifications you mentioned. However, I do not believe that is the correct behavior (or better said not for all the needed cases). From my understanding of the semantic for OVER - even if the function would work on the time (proctime/eventtime) - we should still emit a value for every incoming event. I have 2 arguments for this:
1) in our scenarios - we would use the current implementation for example to detect certain statistics for every incoming event, while the statistic focus is defined for a certain period of time (this is a functionality that is highly needed). For example if you apply this in a stock market scenario - you might want to say give me the sum of the transactions over the last hour (to verify potentially a threshold for liquidity of the market) and as the application would need to react on each incoming transaction (e.g. decide to buy or not to buy) - then working on the behavior you mentioned would not enable such a scenario. More than this, even if you would need both behaviors ..then what query could you write to have the described behavior and make the differentiation from the other?
2) if you think on the case of event time - which should be similar with proctime - then there it should be the same. When you get an event (ev1 , time1) - you should not emit this output until you would know if there is at some point later another event with the same event time (ev2, time1). Basically you would register the timer for the acceptable watermark/allowedlatency and accumulate the accumulator for a specific event time and emit it after the allowed latency has passed....is this the actual behavior that is implemented / would be implemented?

rtudoran commented:
@fhueske  - you can check also the definition from the Calcite website
https://calcite.apache.org/docs/stream.html

Standard SQL features so-called “analytic functions” that can be used in the SELECT clause. Unlike GROUP BY, these do not collapse records. For each record that goes in, one record comes out. But the aggregate function is based on a window of many rows.

fhueske commented:
Sorry, I think did not explain the change correctly. 
I did not mean to group multiple record together. 
We still emit one row per input row, but given the following input

```
(proc-time, id, a)
(1, a, 1)
(2, b, 2)
(3, c, 3)
(3, d, 4)
(4, e, 5)
```
and the query 
```
SELECT id, sum(a) OVER (ORDER BY proctime RANGE BETWEEN 2 PRECEDING AND CURRENT ROW) FROM x
```

the result would not be
```
(a, 1), (b, 3), (c, 6), (d, 10), (e, 14)
```
and not 
```
(a, 1), (b, 3), (c, 10), (d, 10), (e, 14)
```

Because `c` and `d` arrived at the same time, their aggregations need to be the same. 

This is actually also the implementation for event-time OVER RANGE windows.

rtudoran commented:
@fhueske Thanks for the example. First of all i am happy that we agree that we need to emit something for every input :). I was scared that this will not be the case

Now regarding the 2 options for the output - i would still have a preference for the first option. The reason for this is that at least in the case of proctime there is no 2 events that are simultaneous in absolute terms. There is an implicit serialization (also based on how flink engine works...one event will arrive before the other). It might simply be a coincidence that the granularity that the computers now cannot measure system time with more fine granularity...and because of that we would have 2 events which are apparently with the same proc time stamp...
However, as you mentioned several times (and i agree) having a uniform behavior might be a key point. In this case we would indeed implement the behavior you suggested. - Let me know again if you are sure about this. I raised my concerns but i am open to accept what you suggest.

In case you suggest to implement it with proctimes, i have 2 questions:
1) do you know if there is some example for timer in proc time (if not, now problem -  i will figure it out)
2) in the case of harness tests - the correct implementation of the behavior might not match the test. If i write a test that does
{code}
 testHarness.setProcessingTime(3)
 testHarness.processElement(new StreamRecord(rInput, 1001))  // let us assume that the system time is 1490906681
 testHarness.processElement(new StreamRecord(rInput, 2002))  // let us assume that the system time is still 1490906681
 +    testHarness.processElement(new StreamRecord(rInput, 2003))  // let us assume that the system time is now 1490906682
 +    testHarness.processElement(new StreamRecord(rInput, 2004)) // let us assume that the system time is now 1490906682
{code}

in this case if we do a Count, than the output would be
ev1 2
ev2 2
ev3 2
ev4 2
...instead of having all with an associated count of 4 (which would be in the behavior you mention). This would be because i would assume that the timer +1ms would be triggered based on system time.
In such a case - should we just build a limited test that would work?

rtudoran commented:
ignore question 1 - i found an example in the UnboundedEventTimeOverProcessFunction.scala

fhueske commented:
You are right, with processing time no two records arrive at exactly the same time. However, we also retract values based on the granularity of milliseconds. Therefore, I think is makes sense to also do the accumulation on this granularity. The other reason for the change would be consistency with event-time as you mentioned.

The `ProcessFunction` does also support to register processing time timers via the context object.

With the harness we decouple the processing time from the system time. I would hope that the harness also triggers a timer when we increment the processing time but I'm not 100% sure it does.

rtudoran commented:
@fhueske  OK - i will implement this and ping you when it is ready.
One more question - when you say:
"The ProcessFunction does also support to register processing time timers via the context object."
i guess you refer to something like using instead of
 ctx.timerService.registerEventTimeTimer(curWatermark + 1)

 ctx.timerService.registerProcTimeTimer(curProcTime + 1)

rtudoran commented:
for the harness tests - i would let you know after it is implemented...in the case it does not work i would temporary change the tests - that is until a new version of harness would be implemented/fixed

rtudoran commented:
@fhueske 
I just thought of an example. Assume you have 2 events coming at consecutime processintimes
Ev1,1
Ev2, 2
When ev1 arrives you accumulate it and register the timer at proctime +1 to emit it. However when ontimer would be called also the processelement is called. This might lead to some concurrent access for accessing the accumulator as one function needs to read it while the other needs to modify it. Is there a clear order for the execution in such concurrent scenarios?

fhueske commented:
Hi @rtudoran, 
the calls to `processElement()` and `onTimer()` and guarded by a lock. So both methods won't be called at that the same time.

rtudoran commented:
I found a workaround this issue - i will accumulate events in processElement and compute and emit in onTImer

rtudoran commented:
@fhueske it is good to know, but i think we can optimize the computation as mentioned above

fhueske commented:
you're right. That works as well. :-)

rtudoran commented:
@fhueske Thanks for the feedback.
It has on the one hand that the few events that arrive with the same proctime are not computed iteratively, but in a batch (all in the same time - 1 ms later), but it avoids all the mess of having strong synchronization in case there is no explicit and fixed order of execution between the 2 functions (as per my example in the mailing list)

rtudoran commented:
@fhueske  the problem now is that when we run the ITCase tests they are quite random due to the different latency that might appear in the processing time. I think this are quite unreliable and we should not rely on them

fhueske commented:
I agree, let's drop the ITCases for that. The test harness tests should be sufficient.

rtudoran commented:
@fhueske  Great - i was just randomly searching for cases to be valid in single unit runs/cluster run/ide run ...which apparently have widely different latencies :))

rtudoran commented:
@fhueske 

Based on the discussions on the PR this commit includes:
-an implementation based on onTimer in which multiple events arriving in the same proctime would get the same accumulator values
-a removal of the ITCase tests
-a consolidation of the tests done based on the harness framework for proctime

rtudoran commented:
@fhueske  for some reason there was some strange conflict in the SQLITCase  class. I marked is as read. As this final commit does not contain any tests anymroe in the SQLITCase class, when you integrate this PR you can just ignore the class.

fhueske commented:
Thanks for the update @rtudoran.
I will clean that up and merge this PR.

Thanks, Fabian

narendraj9 commented:
@fhueske Hi, I am trying to understand how Objects of RichProcessFunction class are managed. For example, you mentioned that onTimer() and processElement() won't be called at the same time. Could you share the document/link or code that I can read?

fhueske commented:
This might answer some questions: https://ci.apache.org/projects/flink/flink-docs-release-1.4/dev/stream/operators/process_function.html

narendraj9 commented:
@fhueske Thanks for the link to documentation. I have gone through the document. Can you also share a link to the class that would manage a RichProcessFunction instance and make sure that onTimer() and onProcessElement() calls are not interleaved.
